File: 백준/Gold/14500. 테트로미노/테트로미노.py
import sys
from collections import deque 
input = sys.stdin.readline

dx = [-1,1,0,0]
dy = [0,0,-1,1]

def dfs(x,y,cnt,result,visited):
  global max_result
  if(cnt > 1):
    for i in range(4):
      nx = x + dx[i]
      ny = y + dy[i]
      if(0<=nx<n and 0<=ny<m):
        if(nx, ny) not in visited:
          dfs(nx, ny, cnt - 1, result + data[nx][ny], visited + [(nx,ny)])
          dfs(x, y, cnt - 1, result + data[nx][ny], visited + [(nx,ny)])
  else:
    # print()
    # tmp = [[0]*m for _ in range(n)]
    # for a,b in visited:
    #   tmp[a][b] = 1
    # for i in range(n):
    #   print(tmp[i])
    # print(result)
    max_result = max(max_result, result)


n,m = map(int, input().strip().split())
data = []
for _ in range(n):
  data.append(list(map(int, input().strip().split())))


max_result = 0
for i in range(n):
  for j in range(m):
    dfs(i,j,4,data[i][j],[(i,j)])
    
print(max_result)

File: 백준/Gold/14500. 테트로미노/테트로미노.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.*;

public class Main {
    static int n, m;
    static int max_value = 0;
    static List<Integer>[] data;

    static int[] dx = {-1, 1, 0, 0};
    static int[] dy = {0, 0, -1, 1};

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        String[] strings = br.readLine().split(" ");
        n = Integer.parseInt(strings[0]);
        m = Integer.parseInt(strings[1]);


        data = new List[n];


        for(int i = 0; i<n; i++){
            data[i] = new ArrayList<Integer>(m);
            strings = br.readLine().split(" ");
            for(int j = 0; j<m; j++){
                data[i].add(Integer.parseInt(strings[j]));
            }
        }


        for(int i = 0; i<n; i++){
            for(int j = 0; j<m; j++){
                ArrayList<Point> points = new ArrayList<>();
                points.add(new Point(i, j));

                dfs(3, i, j, data[i].get(j), points);
            }
        }

        System.out.println(max_value);
    }

    static void dfs(int cnt, int x, int y, int sum, ArrayList<Point> points){
        if(cnt > 0) {
            for (int i = 0; i < 4; i++) {
                int nx = x + dx[i];
                int ny = y + dy[i];

                if(0 <= nx && nx < n && 0 <= ny && ny < m){

                    Point newPoint = new Point(nx, ny);
                    if(!points.contains(newPoint)){
                        ArrayList<Point> newPoints = new ArrayList<>();
                        Iterator<Point> it = points.iterator();
                        while(it.hasNext()){
                            newPoints.add(it.next());
                        }

                        newPoints.add(newPoint);
                        dfs(cnt - 1, x, y, sum + data[nx].get(ny), newPoints);
                        dfs(cnt - 1, nx, ny, sum + data[nx].get(ny), newPoints);
                    }
                }
            }
        } else{

            max_value = Math.max(sum, max_value);
        }
    }
    
    static class Point{
        int x;
        int y;

        public Point(int x, int y){
            this.x = x;
            this.y = y;
        }

        @Override
        public boolean equals(Object o) {

            if(o != null && o instanceof Point) {
                Point p = (Point)o;
                if(p.x == this.x && p.y == this.y) {
                    return true;
                }
                return false;
            }
            return false;
        }
    }
}